### DISCUSSION TITLE
How does context in relation to django-components work?

### DISCUSSION THREAD

**Discussion by timothyis** [django-components/django-components#226] [Q&A] [answered]
When using a `component_block` and filling a slot with HTML in a template, if I use a `{{ variable }}` within that slot, but use `only` on the block definition, the variable then doesn't fill. 

I assume that this is because then the component_block doesn't have the context of the template within the slot. But this feels like it should be possible. I wonder if this is a limitation of the framework, rather than the package? 

I really like enforcing the context that the component receives with parameters, but when it's a slot, it feels as if that's part of the template the block is placed in and should have the context, even when using `only`?

**Comment by EmilStenstrom** [comment]
@timothyis Thanks for the bug (I think) report. Could you show some sample code of how you use "only"? I don't think we have any tests related to the usage of only, so it's no surprise that this doesn't work as expected.

**Reply by timothyis** [reply]
Yes, of course.

So for example, we have a menu component that has a "trigger" slot for a button that triggers the dropdown menu, and then within an "items" slot, we put menu items. However, the trigger slot is optional and there's a button in the slot by default that receives arguments from the `component_block` definition to fill out the label, if it has an icon, etc.:

```py
Hello {{ username }}

{% component_block label="Open Menu" iconName="placeholder" only %}
	{% slot "items" %}
		<a href="/">Home</a>
		<a href="/logout">Logout {{ username }}</a>
	{% endslot}
{% endcomponent_block %}
```

In this example, we're offering a user to logout and displaying the username. However, this will not load the username, which exists in the templates (the first username, outside the `component_block` will render) because `only` restricts the component itself from receiving the full template context, which we don't want in full.

What I'd expect is that the context shouldn't be passed to the component template itself (the actual menu HTML), but within the slot, because we're defining that in a view template, it should have access to that context. I assume it doesn't because the slot is considered as part of the component template when it renders.

**Reply by timothyis** [reply]
Not sure if this is a bug or intended, by the way. If it's intended that's fine, but at the moment I'm passing entire page context to a template, just for a username or small information like that, that feels like it should be available because the HTML is defined in the view template.

**Reply by EmilStenstrom** [reply]
Hmm. I think I would not expect username not to be available inside the slot. I think of only as a firewall between everything outside of the block, and everything inside of the block. This means that in the example above, if you changed it to:

```django
Hello {{ username }}
{% component_block "menu" username=username only %}
    {% slot "items" %}
        Logout {{ username }}
    {% endslot %}
{% end_component_block %}
```

It should work. But only was added a long time ago, and not by me, so I'm really not sure what the right way is, or even if my expectation above works. Would you mind trying it out?

**Reply by timothyis** [reply]
Yes, that definitely works, but my problem is that `menu` is incredibly diverse and not used in one place where we can say what we need to pass to it. It's the same with other components, for example we have a tooltip that accepts HTML using slots, and a dialog modal component as well that fills the body content with a slot. It would be hard to define everything we'd need over time in these components for every place they're in, e.g., every action of every dialog would have to be a property of the dialog component.

**Reply by timothyis** [reply]
Please don't stress over it too much though, I just thought I'd bring it up since you mentioned in my other issue that you'd like to support common use0cases and I though this was an interesting paradigm of usage that I feel front-end devs like myself are used to from React. But I do understand that this isn't React, so not everything will be like that 😅

**Reply by EmilStenstrom** [reply]
I still confused as to the use-case I think. In the example code you posted, what is the point of using only there? What context are you trying to avoic leaking into the component, and what context are you looking to allow inside it?

**Reply by timothyis** [reply]
Apologies for the late reply. What I'm trying to say is that I don't want the component internals to have access to the context of the page, but I do want the slots that I'm filling to have access to the context. At the moment it's one or the other.

```
{% component_block "menu" only %}
    {% slot "items" %}
        Logout {{ username }}
    {% endslot %}
{% end_component_block %}
```

Taking this as an example, the component itself (the code in the component files) will have access to the context if `only` is omitted here, but so will whatever I'm filling the slots with when I'm using the components on a page.

What I think would be good is if the items slot is still able to access the page context but the component itself aren't able to access the context. 

Since the slot is being filled on the page, maybe it makes sense to allow the slow to have access to the page context, if only is there?

I'm not quite sure how context works and the reasons for it though, so maybe I'm off base here. However, I have very generic components that I can't create props like "username" for, since they might not be used anywhere else.

**Reply by EmilStenstrom** [reply]
I think we have a different idea of what only should be for. It's taken from the [built-in include tag](https://docs.djangoproject.com/en/4.1/ref/templates/builtins/#include). The idea with that one is so that everything inside that include should only have access to the the specific variables sent to it, not anything else. So it's creating an unpassable boundary between this tag and anything outside it. Allowing username to pass through to the slots would break that boundary. So I simply think I don't agree with what only should do in this context.

**Reply by timothyis** [reply]
Yeah, makes sense. I really don't know a lot about Python/Django so I'm just spewing ideas that may help the project since it's made Django 1000% more enjoyable to use when coming from React.

**Comment by JuroOravec** [comment, marked as answer]
For anyone who comes across this thread, this has now been revamped in v0.70 (preferably use v0.72 or higher)

We now support both of the approaches above, as described by @timothyis and @EmilStenstrom.

For @timothyis's approach, in Django settings, set:
```py
COMPONENTS = {
	"context_behavior": "isolated",
}
```

For @EmilStenstrom's approach, in Django settings, set:
```py
COMPONENTS = {
	"context_behavior": "django",
}
```

For more info, see https://github.com/EmilStenstrom/django-components/blob/085c60a8c96442d9fd4f72d66a7eb295e7cfb2d9/README.md#context-behavior
